### PR TITLE
Import Rex::Ui::Text::Table functionality

### PR DESCRIPTION
**Pull request by sempervictus**
Squash historical commits and merge/updates:

---

Allow passing MaxChar to Rex::Ui::Text::Table cols

Passing MaxChar allows setting the maximum number of characters
printed within a specific column during the row_to_s method.
This does not affect CSV output nor truncate the actual data.
Meant for tidying up long console ouput.

Example: cleaned up 'cmd_creds' to show proof and not maul tables
with unix session data.

---

Add MaxChar to tbl.row.cell.to_s

ColProps will now take a 'col_name' => { 'MaxChar' => 64 } which
only affects row.to_s.
cmd_creds without -S will now limit pass and proof to 64 char.

---

Create Rex Tables from CSV

---

Rex::Text::Ui::Table.new[find_by_colnames]

Add :[] to ...Ui::Table allowing user to pass multiple colnames.
Returns a new table with only those columns and their rows.

Useful when using Rex to filter output, prep CSV, etc.

Testing:

```
t = Rex::Ui::Text::Table.new('Columns' => ['a','b','c'])
t << ['x','y','z']
t << ['p','q','r']
t['a','c']

=> a  c
-  -
p  r
x  z
```

---

Allow Internal Filtering by SearchTerm

Allow passing 'SearchTerm' into Rex::Ui::Text::Table creation to
filter all output by regex match to the string passed.
Provides base functionality for higher level subscribers such as
cmd_ls in meterpreter sessions for filtering output

---

Import fs.rb

Merge forked changes to cmd_ls allowing for the use of string
matching on listing output via Rex::Ui::Text::Table's SearchTerm
facility.

Example:

```
meterpreter > ls chef -R -S wget
No entries exist in chef/backup/chef/handlers
No entries exist in chef/backup/chef/ohai_plugins
No entries exist in chef/backup/chef
No entries exist in chef/backup
No entries exist in chef/cache/cookbooks/avast/attributes
No entries exist in chef/cache/cookbooks/avast/recipes
No entries exist in chef/cache/cookbooks/avast
No entries exist in chef/cache/cookbooks/chef-client/attributes
No entries exist in chef/cache/cookbooks/chef-client/libraries
No entries exist in chef/cache/cookbooks/chef-client/recipes
No entries exist in chef/cache/cookbooks/chef-client
No entries exist in chef/cache/cookbooks/chef_handler/attributes
No entries exist in chef/cache/cookbooks/chef_handler/libraries
No entries exist in chef/cache/cookbooks/chef_handler/providers
No entries exist in chef/cache/cookbooks/chef_handler/recipes
No entries exist in chef/cache/cookbooks/chef_handler/resources
No entries exist in chef/cache/cookbooks/chef_handler
No entries exist in chef/cache/cookbooks/cron/providers
No entries exist in chef/cache/cookbooks/cron/recipes
No entries exist in chef/cache/cookbooks/cron/resources
No entries exist in chef/cache/cookbooks/cron
No entries exist in chef/cache/cookbooks/logrotate/attributes
No entries exist in chef/cache/cookbooks/logrotate/definitions
No entries exist in chef/cache/cookbooks/logrotate/libraries
No entries exist in chef/cache/cookbooks/logrotate/recipes
No entries exist in chef/cache/cookbooks/logrotate
No entries exist in chef/cache/cookbooks/ohai/attributes
No entries exist in chef/cache/cookbooks/ohai/files/default/plugins
No entries exist in chef/cache/cookbooks/ohai/files/default
No entries exist in chef/cache/cookbooks/ohai/files
No entries exist in chef/cache/cookbooks/ohai/recipes
No entries exist in chef/cache/cookbooks/ohai
No entries exist in chef/cache/cookbooks/svit-windows/attributes
No entries exist in chef/cache/cookbooks/svit-windows/recipes
No entries exist in chef/cache/cookbooks/svit-windows/templates/default/plugins
No entries exist in chef/cache/cookbooks/svit-windows/templates/default
No entries exist in chef/cache/cookbooks/svit-windows/templates
No entries exist in chef/cache/cookbooks/svit-windows
No entries exist in chef/cache/cookbooks/windows/attributes
No entries exist in chef/cache/cookbooks/windows/files/default/handlers
No entries exist in chef/cache/cookbooks/windows/files/default
No entries exist in chef/cache/cookbooks/windows/files
No entries exist in chef/cache/cookbooks/windows/libraries
No entries exist in chef/cache/cookbooks/windows/providers
No entries exist in chef/cache/cookbooks/windows/recipes
No entries exist in chef/cache/cookbooks/windows/resources
No entries exist in chef/cache/cookbooks/windows
No entries exist in chef/cache/cookbooks
No entries exist in chef/cache
No entries exist in chef/handlers
No entries exist in chef/log
No entries exist in chef/ohai_plugins
No entries exist in chef/run
Listing: chef
=============

Mode              Size  Type  Last modified              Name
----              ----  ----  -------------              ----
100666/rw-rw-rw-  161   fil   2014-07-21 11:08:26 -0400  wget.ps1
100666/rw-rw-rw-  1285  fil   2014-07-21 11:08:26 -0400  wget.vbs

meterpreter >
```

---

Import sys.rb

Merge forked changes to cmd_ps allowing for the use of string
matching on listing output via Rex::Ui::Text::Table's SearchTerm
facility

Example:

```
meterpreter > ps -S x64.*Auth.*Sys

Process list
============

 PID   Name                       Arch  Session  User                          Path
 ---   ----                       ----  -------  ----                          ----
 400   smss.exe                   x64   0        NT AUTHORITY\SYSTEM           C:\Windows\System32\smss.exe
...
```

### CODE REVIEW

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb [hunk 1]
```diff
@@ -30,6 +30,19 @@ class Console::CommandDispatcher::Stdapi::Fs
...
     "-h" => [ false, "Help banner." ],
     "-r" => [ false, "Upload recursively." ])
+  #
+  # Options for the ls command
+  #
+  @@ls_opts = Rex::Parser::Arguments.new(
+    "-h" => [ false, "Help banner." ],
+    "-S" => [ true, "Search string." ],
+    "-t" => [ false, "Sort by time" ],
+    "-t" => [ false, "Sort by time" ],
```
Comment (void-in) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb]: Duplicate -t option. One should be removed.

File: lib/rex/ui/text/table.rb [hunk 2]
```diff
@@ -274,14 +323,17 @@ def columns_to_s # :nodoc:
...
-          remainder = 0
-        end
+        # This produces clean to_s output without truncation
+        # Preserves full string in cells for to_csv output
+        padding = pad(' ', last_col, last_idx)
+        nameline << padding
+        remainder = padding.length - cellpad
+      if (remainder < 0)
+        remainder = 0
+      end
```
Comment (void-in) [inline lib/rex/ui/text/table.rb]: Ternary operator will be compact here
```ruby
rem = padding.length - cellpad
remainder = rem > 0 ? rem : 0
```

File: lib/rex/ui/text/table.rb [hunk 3]
```diff
@@ -330,13 +382,17 @@ def row_to_s(row) # :nodoc:
   # some text and a column index.
   #
   def pad(chr, buf, colidx, use_cell_pad = true) # :nodoc:
-    remainder = colprops[colidx]['MaxWidth'] - buf.length
-    val       = chr * remainder;
+    # Ensure we pad the minimum required amount
+    max = colprops[colidx]['MaxChar'] || colprops[colidx]['MaxWidth']
+    max = colprops[colidx]['MaxWidth'] if max.to_i > colprops[colidx]['MaxWidth'].to_i
+    remainder = max - buf.length
+    remainder = 0 if remainder < 0
```
Comment (void-in) [inline lib/rex/ui/text/table.rb]: Ternary operator can be used here as well.

**Comment by bcook-r7** [comment]
This changes table alignment in places, causing the spec failure. Is this expected? If so, the spec needs to be updated.

**Comment by bcook-r7** [comment]
On converting ls to use Rex::Parser::Arguments, I have mixed feelings. I realize that it is canon and the usual style in a lot of these files. But Rex::Parser::Arguments feels like an anti-pattern to me, just copying the verbosity of C's getopt_long.

In this example, we went from 32 to 52 lines of code to do the exact same parsing task. With a dynamic language like ruby, it would be nice if the library code did more.

I'm fine changing it in the name of consistency, but might prefer if we addressed it in a separate PR from one that also reworks usage of Rex tables.

**Comment by sempervictus** [comment]
Thanks for taking a look @bcook-r7.
The table alignment changes are intended, though i did not update spec on my end since this code is actually several years old (might even be in a closed PR from the past). I and a some of my team have been using msfconsole this way for a couple of years - it helps us to format things like long creds listings and custom outputs (this is also heavily used in our WMI plugin to format annoying WMI output).
Far as the opt parsing bit goes, its not the "same parsing task" - the opt parser, which is standard to framework, is handling a parameter to the -S switch, which the "clean" options mechanism for cmd_ls cannot do. There's more than cmd_ls and cmd_ps that use the table filters in our framework fork, and -S is actually standard for search here (thus always taking a param and sticking to the conventional parser). Also, lines of code within a parsed language tree is a bit of a strange metric - Ruby is constantly extending the actual code processed, and in our case, we do obnoxious amounts of damage to the VM's object cache and GC with our extensions of composed objects after they've been registered with the GC (try it on rubinius and profile what those .extend calls do). In fact, the multiple calls to delete arguments from the array passed in here are actually more abusive than the block in .parse because they hit the GC in MRI with every locking access which iterates through the array to remove an object.
I agree in principle about style complexity in the parser, but do not believe it warrants a departure from codified standards present in the framework for years without a plan for converting the entire codebase or at least having decided on a new standard prior - this one-off opt parsing system does more damage in terms of consistency than it fixes in terms of style (which is a debatable point anyway given the lack of capacity for taking switch params).
The primary purpose for getting this into a PR is that we keep hitting conflicts with upstream and our future submissions will likely reference this or have to be cleaned up like we've done in years past (which has caused problems getting code merged).
I agree that OptionParser is a cleaner mechanism, but we should probably figure out how we want to proceed across the board with that if we're to start making changes on the fly. There's also plugins to consider - i'm sure we're not the only folks here with a small mountain of in-house code in the form of plugins and hackery, and we should at least keep the current parser for backwards compat.

File: lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb (hunk 1 above)
Comment (sempervictus) [inline lib/rex/post/meterpreter/ui/console/command_dispatcher/stdapi/fs.rb]: Nice catch, thanks.

File: lib/rex/ui/text/table.rb (hunk 2 above)
Comment (sempervictus) [inline lib/rex/ui/text/table.rb]: Agreed, not sure why this is formatted as such, believe i got some flak back in the day for getting too "clever" with ternary ops and other syntactic sugar which made my submissions look very inconsistent with upstream. Will update

File: lib/rex/ui/text/table.rb (hunk 3 above)
Comment (sempervictus) [inline lib/rex/ui/text/table.rb]: Agree, but this reads cleaner IMO - Ruby newcomers may have some difficulty with ternary operators (keep seeing it confuse new devs like asking them to find keys to the Humvee). Can change it if there's common desire of course.

**Comment by bcook-r7** [comment]
Fair enough. Standardizing argument parsing for the sake of consistency is more than enough of an argument, without going into the depths of garbage collector performance :) I'd still prefer simpler code eventually, especially in performance-irrelevant paths like user interactive command line parsing. But we digress.

So, the general things we need to land this are, fix the spec and the current code comments. There is also a lot of unrelated code in a single commit here. Squashing is fine in moderation, but this code can now no longer be bisected. It should be split into perhaps 4 logical commits:

  - whitespace formatting changes
  - convert ls argument parsing to use Rex
  - table class improvements and conversions
  - add ls -S argument

For ls,  you think that '-S' is really needed, or could we automatically trigger on the combination of '-r' and the regular argument looking like a pattern (is_glob?) ? It might be more of a principal of least surprise thing.

If you don't mind, I can certainly take on the task of splitting and fixing this up tomorrow. If you want to do it yourself, the tools I use for splitting commits are:

git reset HEAD~  (uncommits the current big commit)
git add --patch / git commit --patch (this starts an interactive mode for cherry-picking bits into logical commits - this is _awesome_)

Then, I use this alias to separate whitespace commits from non-whitespace commits:
whitespace = !sh -c 'git diff -w --no-color "$@" | git apply --cached --ignore-whitespace' -

**Comment by sempervictus** [comment]
@bcook-r7: Please feel free to cut up and reorder as you see fit. My apologies for the lag, juggling chainsaw wielding pumas a fair bit these days. I'll be glad to test/review any results and dive this a bit over the weekend.

**Comment by bcook-r7** [comment]
Thanks. I did some initial slicing and dicing today. I'm going to check with upstream consumers next week to see what other tests might break with table format changes. We can probably land this in a couple of merges. I like the ps regex :)

**Comment by sempervictus** [comment]
Just did a quick look through my framework and it looks like i have the -S flag on most cmd_* methods in lib/msf/ui/console/command_dispatcher/core.rb, the sys, fs, and net public extensions. 
The regex search is also pretty useful for netstat commands since we're usually looking for a specific link, not the dumped table, and marshalling tables directly to CSV is pretty useful for those of us who jump to pry/edit half the time anyway. 
Regarding alignment, there's a method somewhere in here for rotating tables 90 degrees to the left - creating a columnar structure for output, since things like WMI attributes end up being way too wide for a standard console, and the tables get mauled on output. We might want to try that for creds and details of web pages/vulns if this passes muster.

**Comment by bcook-r7** [comment]
@sempervictus here is the resplit version, with some fixes, namely with an empty column in the middle causing the rightmost column to be misaligned.

https://github.com/bcook-r7/metasploit-framework/commits/land-5341-table-others

**Comment by bcook-r7** [comment]
Landed, though I need to close this manually, since I split up the first commit. Thanks for the improvements @sempervictus!

**Pull request closed without merging**